File: gtsam-rs/sys/src/slam/projection_factor.cc
#include "projection_factor.h"
#include "../base/rust.hpp"

namespace gtsam
{
    std::shared_ptr<SmartProjectionPoseFactorCal3_S2> new_smart_projection_pose_factor(
        const std::shared_ptr<noiseModel::Isotropic> & measurement_noise,
        const std::shared_ptr<Cal3_S2> & K,
        const Pose3 & sensor_P_body)
    {
        SmartProjectionParams params;
        params.setRankTolerance(1);
        params.setLandmarkDistanceThreshold(10);
        params.setRetriangulationThreshold(0.001);
        params.setDynamicOutlierRejectionThreshold(3);
        //! EPI: If set to true, will refine triangulation using LM.
        params.setEnableEPI(false);
        params.setLinearizationMode(gtsam::HESSIAN);
        params.setDegeneracyMode(gtsam::ZERO_ON_DEGENERACY);
        params.throwCheirality = false;
        params.verboseCheirality = false;

        auto factor = SmartProjectionPoseFactor<gtsam::Cal3_S2>(
            to_boost_ptr(measurement_noise),
            to_boost_ptr(K),
            sensor_P_body,
            params
        );

        return std::shared_ptr<SmartProjectionPoseFactor<gtsam::Cal3_S2>>(new SmartProjectionPoseFactor<gtsam::Cal3_S2>(factor));
    }

    void add_smart(
        std::shared_ptr<SmartProjectionPoseFactorCal3_S2> &factor,
        const Point2 & point,
        Key key
    ) {
        factor->add(point, key);
    }

    std::shared_ptr<SmartProjectionPoseFactorCal3_S2> clone_smart_projection_pose_factor(
        const std::shared_ptr<SmartProjectionPoseFactorCal3_S2> & old_factor
    ) {
        return std::shared_ptr<SmartProjectionPoseFactorCal3_S2>(new SmartProjectionPoseFactorCal3_S2(*old_factor));
    }

    std::shared_ptr<gtsam::SmartStereoProjectionPoseFactor> new_smart_stereo_projection_pose_factor(
        const std::shared_ptr<noiseModel::Base> &measurement_noise,
        const Pose3 & sensor_P_body)
    {
        SmartStereoProjectionParams params;
        params.setRankTolerance(1);
        params.setLandmarkDistanceThreshold(10);
        params.setRetriangulationThreshold(0.001);
        params.setDynamicOutlierRejectionThreshold(3);
        //! EPI: If set to true, will refine triangulation using LM.
        params.setEnableEPI(false);
        params.setLinearizationMode(gtsam::HESSIAN);
        params.setDegeneracyMode(gtsam::ZERO_ON_DEGENERACY);
        params.throwCheirality = false;
        params.verboseCheirality = false;

        // smart_noise_, smart_factors_params_, B_Pose_leftCamRect_

        return std::shared_ptr<gtsam::SmartStereoProjectionPoseFactor>(
            new gtsam::SmartStereoProjectionPoseFactor(
                to_boost_ptr(measurement_noise),
                params,
                sensor_P_body)
        );
    }

    std::shared_ptr<gtsam::SmartStereoProjectionPoseFactor> clone_smart_stereo_projection_pose_factor(
        const std::shared_ptr<gtsam::SmartStereoProjectionPoseFactor> & old_factor
    ) {
        return std::shared_ptr<gtsam::SmartStereoProjectionPoseFactor>(new gtsam::SmartStereoProjectionPoseFactor(*old_factor));
    }

    void add_smartstereo(
        std::shared_ptr<gtsam::SmartStereoProjectionPoseFactor> &factor,
        const StereoPoint2 &point,
        Key key,
        const std::shared_ptr<Cal3_S2Stereo> &K
    ) {
        factor->add(point, key, to_boost_ptr(K));
    }

    std::shared_ptr<gtsam::GenericProjectionFactor<Pose3, Point3, Cal3_S2>> new_generic_projection_factor(
        const Point2 & measured,
        const std::shared_ptr<noiseModel::Isotropic> & measurement_noise,
        Key pose_key,
        Key point_key,
        const std::shared_ptr<Cal3_S2> & K,
        const Pose3 & sensor_P_body)
    {
        auto factor = gtsam::GenericProjectionFactor<Pose3, Point3, Cal3_S2>(
            measured,
            to_boost_ptr(measurement_noise),
            pose_key,
            point_key,
            to_boost_ptr(K),
            sensor_P_body);

        return std::shared_ptr<gtsam::GenericProjectionFactor<Pose3, Point3, Cal3_S2>>(new gtsam::GenericProjectionFactor<Pose3, Point3, Cal3_S2>(factor));
    }

} // namespace gtsam

File: gtsam-rs/src/slam/projection_factor.rs
use cxx::SharedPtr;

use crate::{geometry::{cal3_s2::{Cal3S2, Cal3S2Stereo}, point2::{Point2, StereoPoint2}, pose3::Pose3}, inference::key::IntoKey, linear::noise_model::IsotropicNoiseModel};

pub struct SmartProjectionPoseFactorCal3S2 {
    pub(crate) inner: SharedPtr<::sys::SmartProjectionPoseFactorCal3_S2>,
}

impl SmartProjectionPoseFactorCal3S2 {
    pub fn new(
        measurement_noise: &IsotropicNoiseModel,
        k: &Cal3S2,
        sensor_p_body: &Pose3,
    ) -> Self {
        Self {
            inner: ::sys::new_smart_projection_pose_factor(
                &measurement_noise.inner,
                &k.inner,
                &sensor_p_body.inner,
            ),
        }
    }

    pub fn copy(&self) -> Self {
        Self {
            inner: ::sys::clone_smart_projection_pose_factor(&self.inner)
        }
    }

    pub fn add(
        self: &mut Self,
        point: &Point2,
        key: impl IntoKey
    ) {
        ::sys::add_smart(
            &mut self.inner,
            &point.inner,
            key.into_key(),
        )
    }
}

pub struct SmartStereoProjectionPoseFactor {
    pub(crate) inner: SharedPtr<::sys::SmartStereoProjectionPoseFactor>,
}

impl SmartStereoProjectionPoseFactor {
    pub fn new(
        measurement_noise: &IsotropicNoiseModel,
        sensor_p_body: &Pose3,
    ) -> Self {
        Self {
            inner: ::sys::new_smart_stereo_projection_pose_factor(
                &measurement_noise.to_base_model().inner,
                &sensor_p_body.inner,
            ),
        }
    }

    pub fn copy(&self) -> Self {
        Self {
            inner: ::sys::clone_smart_stereo_projection_pose_factor(&self.inner)
        }
    }

    pub fn add(
        self: &mut Self,
        point: &StereoPoint2,
        key: impl IntoKey,
        k: &Cal3S2Stereo,
    ) {
        ::sys::add_smartstereo(
            &mut self.inner,
            &point.inner,
            key.into_key(),
            &k.inner,
        )
    }
}



pub struct GenericProjectionFactorPose3Point3Cal3S2 {
    pub(crate) inner: SharedPtr<::sys::GenericProjectionFactorPose3Point3Cal3_S2>,
}

impl GenericProjectionFactorPose3Point3Cal3S2 {
    pub fn new(
        point: &Point2,
        measurement_noise: &IsotropicNoiseModel,
        pose_key: impl IntoKey,
        point_key: impl IntoKey,
        k: &Cal3S2,
        sensor_p_body: &Pose3,
    ) -> Self {
        // From gtsam:
            // measured	is the 2 dimensional location of point in image (the measurement)
            // model	is the standard deviation
            // poseKey	is the index of the camera
            // pointKey	is the index of the landmark
            // K	shared pointer to the constant calibration
            // body_P_sensor	is the transform from body to sensor frame (default identity)
        Self {
            inner: ::sys::new_generic_projection_factor(
                &point.inner,
                &measurement_noise.inner,
                pose_key.into_key(),
                point_key.into_key(),
                &k.inner,
                &sensor_p_body.inner,
            ),
        }
    }
}